File: benchmarks/go/05_matrix_mul.go
// Matrix Multiplication Benchmark - Go Implementation
// Multiply two NxN matrices using arrays

package main

import "fmt"

func main() {
	// Benchmark parameter
	n := 50

	// Initialize matrices
	a := make([]int64, n*n)
	b := make([]int64, n*n)
	c := make([]int64, n*n)

	// Fill matrices with values
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			idx := i*n + j
			a[idx] = int64((i + j) % 10)
			b[idx] = int64((i * j) % 10)
		}
	}

	// Matrix multiplication: C = A * B
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			var sum int64 = 0
			for k := 0; k < n; k++ {
				a_idx := i*n + k
				b_idx := k*n + j
				sum += a[a_idx] * b[b_idx]
			}
			c_idx := i*n + j
			c[c_idx] = sum
		}
	}

	// Calculate checksum
	var checksum int64 = 0
	for _, v := range c {
		checksum += v
	}

	fmt.Printf("matrix %dx%d checksum: %d\n", n, n, checksum)
}

File: benchmarks/c/05_matrix_mul.c
// Matrix Multiplication Benchmark - C Implementation
// Multiply two NxN matrices using arrays

#include <stdio.h>

#define N 50

int main() {
    // Initialize matrices
    long a[N * N];
    long b[N * N];
    long c[N * N];
    
    // Fill matrices with values
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            int idx = i * N + j;
            a[idx] = (i + j) % 10;
            b[idx] = (i * j) % 10;
            c[idx] = 0;
        }
    }
    
    // Matrix multiplication: C = A * B
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            long sum = 0;
            for (int k = 0; k < N; k++) {
                int a_idx = i * N + k;
                int b_idx = k * N + j;
                sum += a[a_idx] * b[b_idx];
            }
            int c_idx = i * N + j;
            c[c_idx] = sum;
        }
    }
    
    // Calculate checksum
    long checksum = 0;
    for (int i = 0; i < N * N; i++) {
        checksum += c[i];
    }
    
    printf("matrix %dx%d checksum: %ld\n", N, N, checksum);
    return 0;
}
